src/test/java/uk/gov/ons/ctp/response/notify/service/CommsTemplateClientTest.java

```diff
@@ -0,0 +1,108 @@
+import uk.gov.ons.ctp.common.rest.RestUtilityConfig;
+import uk.gov.ons.ctp.response.notify.client.CommsTemplateClientException;
+import uk.gov.ons.ctp.response.notify.config.AppConfig;
+import uk.gov.ons.ctp.response.notify.config.CommsTemplateService;
+import uk.gov.ons.ctp.response.notify.domain.model.CommsTemplateDTO;
+import uk.gov.ons.ctp.response.notify.service.impl.CommsTemplateClientImpl;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+
+
+import java.io.IOException;
+import java.net.URI;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CommsTemplateClientTest {
+
+    @InjectMocks
+    private CommsTemplateClientImpl commsTemplateClient;
+
+    @Mock
+    private CommsTemplateService commsTemplateService;
+
+    @Mock
+    private AppConfig appConfig;
+
+    @Mock
+    private ObjectMapper objectMapper;
+
+    @Mock
+    private RestTemplate restTemplate;
+
+    @Spy
+    private RestUtility restUtility = new RestUtility(new RestUtilityConfig());
+
+
+    @Before
+    public void setUp() {
+        Mockito.when(appConfig.getCommsTemplateService()).thenReturn(commsTemplateService);
+        Mockito.when(commsTemplateService.getTemplateByClassifiersPath()).thenReturn("/templates");
+    }
+
+    @Test
+    public void testMakesCorrectGetRequest() throws CommsTemplateClientException, IOException {
+        //Given
+        Mockito.when(restTemplate.exchange(any(URI.class), eq(HttpMethod.GET),
+                any(HttpEntity.class), eq(String.class))).thenReturn(new ResponseEntity(HttpStatus.OK));
+
+        //passed in like this as can't mix matchers and concrete values, if not declared as string then
+        // mockito complains it is an ambiguous method call (could be one of 2 methods)
+        String nullString = null;
+        Mockito.when(objectMapper.readValue(nullString, CommsTemplateDTO.class)).thenReturn(CommsTemplateDTO.builder().build());
+
+        ArgumentCaptor<URI> uriArgumentCaptor = ArgumentCaptor.forClass(URI.class);
+
+        //When
+        commsTemplateClient.getCommsTemplateByClassifiers(generateClassifiers());
+
+        //Then
+        Mockito.verify(restTemplate).exchange(uriArgumentCaptor.capture(), eq(HttpMethod.GET),
+                anyObject(), eq(String.class));
+        String expectedURL = "http://localhost:8080/templates?LEGALBASIS=YY&REGION=NI";
+        assertEquals(expectedURL, uriArgumentCaptor.getValue().toString());
+    }
+
+    @Test
+    public void testThrowsExceptionWhenUnableToFindTemplate() {
+        //Given
+        Mockito.when(restTemplate.exchange(any(URI.class), eq(HttpMethod.GET),
+                any(HttpEntity.class), eq(String.class))).thenReturn(new ResponseEntity(HttpStatus.BAD_REQUEST));
+
+        try {
+            //When
+            commsTemplateClient.getCommsTemplateByClassifiers(generateClassifiers());
```

Comment:
If the exception isn't throw the test will still pass

Reply:
I've changed it to receive an ok response and it doesn't pass?

Reply:
I think that's by virtue the logger tries to do something with a null.
```
java.lang.NullPointerException
	at uk.gov.ons.ctp.response.notify.service.impl.CommsTemplateClientImpl.getCommsTemplateByClassifiers(CommsTemplateClientImpl.java:62)
	at uk.gov.ons.ctp.response.notify.service.CommsTemplateClientTest.testThrowsExceptionWhenUnableToFindTemplate(CommsTemplateClientTest.java:92)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.mockito.internal.runners.JUnit45AndHigherRunnerImpl.run(JUnit45AndHigherRunnerImpl.java:37)
	at org.mockito.runners.MockitoJUnitRunner.run(MockitoJUnitRunner.java:62)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Typically you would have a `fail()` after you expect the code to throw an exception which would cause the test to fail if the exception wasn't thrown

```
    @Test
    public void testThrowsExceptionWhenUnableToFindTemplate() {
        //Given
        Mockito.when(restTemplate.exchange(any(URI.class), eq(HttpMethod.GET),
                any(HttpEntity.class), eq(String.class))).thenReturn(new ResponseEntity(HttpStatus.OK));

        try {
            //When
            commsTemplateClient.getCommsTemplateByClassifiers(generateClassifiers());
            fail()
        } catch (CommsTemplateClientException exception) {
            //Then
            assertEquals(CommsTemplateClientException.class, exception.getClass());
            assertEquals(CTPException.Fault.SYSTEM_ERROR, exception.getFault());
        }
    }
```

Reply:
Cheers